### PR TITLE
Use StepListener to Simplify SnapshotResiliencyTests

### PR DESCRIPTION
**Pull request by original-brownbear**
* Reduces complicated callback relations in `testSuccessfulSnapshotAndRestore` to flat steps of sequential actions
  * Note, there are no logic changes to what this test does here, it's really just moving things around
* Would refactor the other tests in this suit as a follow up if others agree that this simplifies things
   * This format would certainly make it easier to create more complicated tests that involve multiple subsequent snapshots as it would allow adding loops

### CODE REVIEW

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/bwc

File: server/src/test/java/org/elasticsearch/snapshots/SnapshotResiliencyTests.java [hunk 1]
```diff
@@ -244,65 +250,71 @@ public void testSuccessfulSnapshotAndRestore() {
...
+            .setType(FsRepository.TYPE).setSettings(Settings.builder().put("location", randomAlphaOfLength(10)))
+            .execute(createRepositoryListener);
+
+        final StepListener<CreateIndexResponse> createIndexResponseStepListener = new StepListener<>();
+        createRepositoryListener.whenComplete(acknowledgedResponse -> masterNode.client.admin().indices().create(
+            new CreateIndexRequest(index).waitForActiveShards(ActiveShardCount.ALL).settings(defaultIndexSettings(shards)),
+            createIndexResponseStepListener), SnapshotResiliencyTests::rethrowAssertion);
+
+        final StepListener<CreateSnapshotResponse> createSnapshotResponseListener = new StepListener<>();
+        createIndexResponseStepListener.whenComplete(createIndexResponse -> {
```
Comment (andrershov) [inline server/src/test/java/org/elasticsearch/snapshots/SnapshotResiliencyTests.java]: 1. If we are using bulk, why not to prepare the single bulk request and send it just once and avoid CountDownLatch usage?
2. After doing it, I would create separate bulkResponseListener and chain it with createSnapshotResponseListener.

File: server/src/test/java/org/elasticsearch/snapshots/SnapshotResiliencyTests.java [hunk 2]
```diff
@@ -244,65 +250,71 @@ public void testSuccessfulSnapshotAndRestore() {
...
+                        }));
+            }
+            if (documents == 0) {
+                afterIndexing.run();
+            }
+        }, SnapshotResiliencyTests::rethrowAssertion);
+
+        final StepListener<AcknowledgedResponse> deleteIndexListener = new StepListener<>();
+
         final AtomicBoolean createdSnapshot = new AtomicBoolean();
```
Comment (andrershov) [inline server/src/test/java/org/elasticsearch/snapshots/SnapshotResiliencyTests.java]: We can avoid this Atomic and Atomics below and use, listener.result() instead

**Review by andrershov — CHANGES_REQUESTED** [review]
@original-brownbear thanks for re-writing this callback hell!
Looks much better now, not as pretty as with Futures and for-comprehensions in Scala, but digestible!
I left 2 comments to make it even better.

File: server/src/test/java/org/elasticsearch/snapshots/SnapshotResiliencyTests.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/test/java/org/elasticsearch/snapshots/SnapshotResiliencyTests.java]: Thanks for spotting, no clue why I made it sent multiple requests here back when I created this :) -> moved to a single bulk request as suggested.

**Comment by original-brownbear** [comment]
@andrershov thanks, fixed the two spots you pointed out now.

File: server/src/test/java/org/elasticsearch/snapshots/SnapshotResiliencyTests.java (hunk 1 above)
Comment (andrershov) [reply server/src/test/java/org/elasticsearch/snapshots/SnapshotResiliencyTests.java]: The first item is addressed, but what about the second item? I suggest introducing new StepListener called bulkResponseListener and chain it createSnapshotResponseListener. This way you can get rid of `CountDownLatch`.

**Review by andrershov — CHANGES_REQUESTED** [review]

File: server/src/test/java/org/elasticsearch/snapshots/SnapshotResiliencyTests.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/test/java/org/elasticsearch/snapshots/SnapshotResiliencyTests.java]: Ah right ... the counter doesn't make much sense now either, killed it and just added an assertion on the count to be safe :)

**Review by andrershov — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
Thanks @andrershov !

**Pull request merged**